soroban-env-host/src/vm/module_cache.rs

```diff
@@ -118,19 +118,13 @@ impl ModuleCache {
         self.with_import_symbols(host, |symbols| Host::make_linker(&self.engine, symbols))
     }
 
-    pub fn get_module(
-        &self,
-        host: &Host,
-        contract_id: &Hash,
-    ) -> Result<Rc<ParsedModule>, HostError> {
-        if let Some(m) = self.modules.get(contract_id, host)? {
+    pub fn get_module(&self, host: &Host, wasm_hash: &Hash) -> Result<Rc<ParsedModule>, HostError> {
+        if let Some(m) = self.modules.get(wasm_hash, host)? {
             return Ok(m.clone());
         } else {
-            Err(host.err(
-                ScErrorType::Context,
-                ScErrorCode::InternalError,
-                "module cache missing contract",
-                &[],
+            Err(host.decorate_contract_code_storage_error(
```

Comment:
What happens if a Wasm key is in the footprint, but the entry itself does not exist? IIUC we would still end up here and return the footprint error, which is not correct. The most comprehensive way to report an error here would probably be to do a storage lookup. If the Wasm is in the footprint, but storage returns `None`, then we can state that non-existent Wasm has been accessed. Otherwise, we can state that an entry outside of the footprint is being accessed. It's a bit of additional metered work, but since it only happens on the error path, I don't think it's an issue.

Reply:
Actually there is yet another scenario, where Wasm could have been uploaded before cache is accessed. @graydon will look into that as well. Probably we should wait with this PR until then, because I'm not sure what the error logic should be.